### PR TITLE
allow labels for gradle wrapper task versions

### PR DESCRIPTION
**Pull request by reinsch82**
permitted labels:
- "latest"
- "release-candidate"
- "nightly"
- "release-nightly"

- Implements: #13504

Signed-off-by: Reinhold Degenfellner <rdegenfellner@gradle.com>

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test AFT

File: subprojects/build-init/src/test/groovy/org/gradle/api/tasks/wrapper/WrapperTest.groovy [hunk 1]
```diff
@@ -18,23 +18,38 @@ package org.gradle.api.tasks.wrapper
...
 import org.gradle.wrapper.WrapperExecutor
 
 class WrapperTest extends AbstractTaskTest {
+    private static final String RELEASE = "7.6"
+    private static final String RELEASE_CANDIDATE = "7.7-rc-5"
+    private static final String NIGHTLY = "8.1-20221207164726+0000"
+    private static final String RELEASE_NIGHTLY = "8.0-20221207164726+0000"
     private Wrapper wrapper
-    private String targetWrapperJarPath
+    private String targetWrapperJarPath = "gradle/wrapper"
```
Comment (eskatos) [inline subprojects/build-init/src/test/groovy/org/gradle/api/tasks/wrapper/WrapperTest.groovy:36]: :lipstick: can be made `final`

**Review by eskatos — COMMENTED** [review]
This is looking good :+1: 

Left a bunch of comments and suggestions. Mostly :lipstick:

**Comment by eskatos** [comment]
@bot-gradle test AFT

**Comment by eskatos** [comment]
@bot-gradle test this

File: subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java [hunk 2]
```diff
@@ -318,10 +308,24 @@ public void setWrapperVersionsResources(WrapperVersionsResources wrapperVersions
...
+    }
+
+    /**
+     * The version of the gradle distribution required by the wrapper.
+     * This is usually the same version of Gradle you use for building your project.
+     * The are labels allowed to specify a version: latest, release-candidate, nightly, and release-nightly
      */
-    @Option(option = "gradle-version", description = "The version of the Gradle distribution required by the wrapper.")
+    @Option(option = "gradle-version", description = "The version of the Gradle distribution required by the wrapper. " +
+        "The following labels are allowed: latest, release-candidate, nightly, and release-nightly")
```
Comment (eskatos) [inline subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java:328]:
```suggestion
        "The following labels are allowed: latest, release-candidate, nightly, and release-nightly.")
```

File: subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc [hunk 3]
```diff
@@ -680,10 +680,11 @@ $ gradle init --type java-library
 The built-in `gradle wrapper` task generates a script, `gradlew`, that invokes a declared version of Gradle, downloading it beforehand if necessary.
 
 ----
-$ gradle wrapper --gradle-version=4.4
+$ gradle wrapper --gradle-version=8.1
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/reference/command_line_interface.adoc:683]: :+1:

File: subprojects/docs/src/docs/userguide/reference/gradle_wrapper.adoc [hunk 4]
```diff
@@ -173,9 +191,15 @@ The Wrapper shell script and batch file reside in the root directory of a single
 [[sec:upgrading_wrapper]]
 == Upgrading the Gradle Wrapper
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/reference/gradle_wrapper.adoc:192]: We should also add some words in this section.

File: subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java [hunk 5]
```diff
@@ -288,23 +293,43 @@ public void setDistributionPath(String distributionPath) {
...
+        gradleVersionResolver.setTextResources(defaultWrapperVersionsResources.getLatest(),
+            defaultWrapperVersionsResources.getReleaseCandidate(),
+            defaultWrapperVersionsResources.getNightly(),
+            defaultWrapperVersionsResources.getReleaseNightly());
+    }
+
     /**
      * Returns the gradle version for the wrapper.
+     * This may throw if the label that can be provided via #setGradleVersion(String) can not be resolved at the moment
+     * e.g. there is not a `release-candidate` available at all times.
```
Comment (eskatos) [inline subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java:313]: :x: These two lines are now redundant with the `@throws` tag and should be removed.

File: subprojects/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -24,6 +24,21 @@ For Java, Groovy, Kotlin and Android compatibility, see the [full compatibility
...
+The [`--gradle-version`](userguide/gradle_wrapper.html#sec:adding_wrapper) parameter for the wrapper plugin 
+now supports using predefined labels to select a version.
+
+The allowed labels are:
+- `latest`
+- `release-candidate`
+- `nightly`
+- `release-nightly`
+
+More Details can be found in the [Gradle Wrapper](userguide/gradle_wrapper.html#sec:adding_wrapper) section.
```
Comment (eskatos) [inline subprojects/docs/src/docs/release/notes.md:40]: :x:
```suggestion
More details can be found in the [Gradle Wrapper](userguide/gradle_wrapper.html#sec:adding_wrapper) section.
```

**Review by eskatos — CHANGES_REQUESTED** [review]
A few tiny changes requested

**Review by eskatos — APPROVED** [review]
Good work!

:tada:

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test and merge

**Comment by reinsch82** [comment]
@bot-gradle test and merge

**Comment by eskatos** [comment]
A follow up issue should be created about supporting `milestone`.

**Comment by eskatos** [comment]
@reinsch82 did you get to create the follow up issue about supporting `milestone`?

We should also review how our custom wrapper tasks in the `gradle/gradle` build could be rewritten using this new feature. See https://github.com/gradle/gradle/blob/master/build-logic/build-update-utils/src/main/kotlin/gradlebuild.wrapper.gradle.kts

**Comment by reinsch82** [comment]
@eskatos just did 
- #23369

should I create an issue for the custom wrapper adaption?

**Comment by eskatos** [comment]
Thanks!

A direct PR changing our build logic would be fine for our custom wrapper tasks.

**Pull request merged**